nbconvert/preprocessors/execute.py

```diff
@@ -466,55 +466,59 @@ def run_cell(self, cell, cell_index=0):
-                display_id = msg['content'].get('transient', {}).get('display_id', None)
-                if display_id:
-                    self._update_display_id(display_id, msg)
-                if msg_type == 'update_display_data':
-                    # update_display_data doesn't get recorded
-                    continue
-
+            if not self.process_message(msg, cell, cell_index):
+                break
+
+        return exec_reply, cell.outputs
+
+    def process_message(self, msg, cell, cell_index):
+        '''
+        Returns None if execution should be halted.
+        '''
+        msg_type = msg['msg_type']
+        self.log.debug("msg_type: %s", msg_type)
+        content = msg['content']
+        self.log.debug("content: %s", content)
+
+        # Default to our input as the "result" of processing the message
+        result = msg
+
+        display_id = content.get('transient', {}).get('display_id', None)
+        if display_id and msg_type in {'execute_result', 'display_data', 'update_display_data'}:
+            self._update_display_id(display_id, msg)
+
+        # set the prompt number for the input and the output
+        if 'execution_count' in content:
+            cell['execution_count'] = content['execution_count']
+
+        if msg_type == 'status':
+            if content['execution_state'] == 'idle':
+                # Set result to None to halt execution
+                result = None
+        elif msg_type == 'clear_output':
+            cell.outputs[:] = []
+            # clear display_id mapping for this cell
+            for display_id, cell_map in self._display_id_map.items():
+                if cell_index in cell_map:
+                    cell_map[cell_index] = []
+        # Check for remaining messages we don't process
+        elif not (msg_type in ['execute_input', 'update_display_data'] or msg_type.startswith('comm')):
             try:
-                out = output_from_msg(msg)
+                # Assign output as our processed "result"
+                result = output_from_msg(msg)
             except ValueError:
                 self.log.error("unhandled iopub msg: " + msg_type)
-                continue
-            if display_id:
-                # record output index in:
-                #   _display_id_map[display_id][cell_idx]
-                cell_map = self._display_id_map.setdefault(display_id, {})
-                output_idx_list = cell_map.setdefault(cell_index, [])
-                output_idx_list.append(len(outs))
+            else:
+                if display_id:
+                    # record output index in:
+                    #   _display_id_map[display_id][cell_idx]
+                    cell_map = self._display_id_map.setdefault(display_id, {})
+                    output_idx_list = cell_map.setdefault(cell_index, [])
+                    output_idx_list.append(len(cell.outputs))
 
-            outs.append(out)
+                cell.outputs.append(result)
```

Comment:
I don't really like the idea of having an API that works primarily via side-effects — that's a difficult API to document and test for. 

Can you think of a way that we could do this as a pure function? E.g., could we adhere to the pattern for we use with preprocessors? I think that would involve returning and reassigning the modified cell object rather than returning the result itself while treating the cell.outputs as a location where we can append changes?

Reply:
I can take a look. Mostly I tried to keep the logic the same and not change the design, just break it into smaller parts.